### PR TITLE
 Extends variant annotation alt matching.

### PR DESCRIPTION
**Pull request by bashir2**
Please note that this is on top of PR #125 and as you can see it has two commits. The first one is the previous PR and can be skipped. So please only review the second commit.

Tested:
Updated unit-tests; ran integration test; manually checked the output table for valid-4.2_VEP.vcf case. For example this query used to return 11 but now returns 26:
`SELECT COUNT(0) FROM (
SELECT start_position, alt.alt, CSQ.Consequence
FROM vcf_test_imports.VEP_result_for_valid_4_2_fixed AS T, T.alternate_bases AS alt, alt.CSQ as CSQ)`
But even with these changes we are not covering all annotation corner cases in gnomAD!

Issue #81

### CODE REVIEW

File: gcp_variant_transforms/libs/processed_variant.py [hunk 1]
```diff
@@ -235,51 +244,18 @@ def _add_annotation(self, proc_var, field_name, data):
...
-    # type: (List[str], List[str]) -> Dict[str, str]
-    if len(annotation_names) != len(annotations) - 1:
-      raise ValueError('Expected {} annotations, got {}'.format(
-          len(annotation_names), len(annotations) - 1))
-    annotation_dict = {}
-    for index, name in enumerate(annotation_names):
-      annotation_dict[name] = annotations[index + 1]
-    return annotation_dict
+    annotation_processor = _AnnotationProcessor(
+        proc_var, field_name, self._annotation_counter,
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/processed_variant.py]: I'm starting to feel uneasy about the mutable proc_var getting passed around like this as it's no longer clear what is being mutated and in what way.
Given that the only mutable part is the alternate base, then does it make sense to redesign this in such a way that you only pass alternate_base? Looks like you also need reference_base for calculating the common prefix, but that can be done outside that class as a 'util' method?

File: gcp_variant_transforms/libs/processed_variant.py [hunk 2]
```diff
@@ -235,51 +244,18 @@ def _add_annotation(self, proc_var, field_name, data):
...
-  def _create_annotation_map(self, annotations, annotation_names):
-    # type: (List[str], List[str]) -> Dict[str, str]
-    if len(annotation_names) != len(annotations) - 1:
-      raise ValueError('Expected {} annotations, got {}'.format(
-          len(annotation_names), len(annotations) - 1))
-    annotation_dict = {}
-    for index, name in enumerate(annotation_names):
-      annotation_dict[name] = annotations[index + 1]
-    return annotation_dict
+    annotation_processor = _AnnotationProcessor(
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/processed_variant.py]: From the API of this class, it's not really clear why some args are in the constructor vs. in the add_annotation_data method (e.g. field_name and annotation_names). It's also not clear why a new instance of the AnnotationProcessor class is needed for every annotation. I think a better design may be to have a single AnnotationProcessor (that has counters) and you'd use that instance for processing any annotation? This paradigm also resolves the issue of why some fields would be in the constructor vs. in the methods.

File: gcp_variant_transforms/libs/processed_variant.py [hunk 3]
```diff
@@ -343,6 +319,101 @@ def info_is_in_alt_bases(self, info_field_name):
...
+    # Handling symbolic ALTs.
+    id_match = _SYMBOLIC_ALT_RE.match(alt_bases)
+    if id_match and id_match.group('ID') == annotation_alt:
+      return True
+    # Handling breakend ALTs.
+    # TODO(bashir2): Check if the following logic is documented anywhere! I
+    # could not find it explicitly in any documentation but that's how I saw
+    # VEP does it in some examples I ran.
+    breakend_match = _BREAKEND_ALT_RE.match(alt_bases)
+    if (breakend_match and breakend_match.group('up_to_chr') == annotation_alt):
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/processed_variant.py]: nit: please remove the surrounding brackets.

File: gcp_variant_transforms/libs/processed_variant.py [hunk 4]
```diff
@@ -343,6 +319,101 @@ def info_is_in_alt_bases(self, info_field_name):
...
+    else:
+      self._mismatch_counter.inc()
+      logging.warning(
+          'Could not find matching alternate bases for %s in '
+          'annotation filed %s', alt_bases, self._anntation_field_name)
+
+  def _alt_matches_annotation_alt(self, alt_bases, annotation_alt):
+    # type: (str, str) -> bool
+    """Returns true if `alt_bases` matches`annotation_alt`
+    See: https://useast.ensembl.org/info/docs/tools/vep/vep_formats.html
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/processed_variant.py]: Please clarify which section in the doc talks about this logic. Also, please summarize the logic here as well (links can become broken over time) and provide a few examples.

**Comment by bashir2** [comment]
To check the changes since your last review, please check [this commit](https://github.com/bashir2/gcp-variant-transforms/commit/1d3fa83f8c338ccfc7768b8004e88d2c6caa7462).

**Comment by bashir2** [comment]
All of the 357 mismatch cases, that I reported for that gnomAD small sample, were covered by implementing what [--minimal option of VEP](https://useast.ensembl.org/info/docs/tools/vep/script/vep_options.html#opt_minimal) does to annotation ALTs. So apparently gnomAD annotation has been done by that feature of VEP which is unfortunate because it can add ambiguous cases (details to come in the next PR which needs schema changes too that I have not done yet). Anyways, a run with no mismatch can be found [here](https://pantheon.corp.google.com/dataflow/jobsDetail/locations/us-central1/jobs/2018-03-03_15_07_06-11713068003808116107?project=gcp-variant-transforms-test&authuser=1&organizationId=433637338589).

**Comment by bashir2** [comment]
PR updated with merging/conflict resolution of changes on `master` and PR #125, plus some more `_AnnotationProcessor` shuffling for better performance.

**Comment by bashir2** [comment]
Just a reminder that this PR (and the next one on top of this) are ready for another round of review for a while now.

Thanks

**Comment by bashir2** [comment]
Added the TODO for the annotation class and merged new changes from master.

Please approve again as the `push -f` overwrote the approval bit; thanks.

**Pull request merged**